### DISCUSSION TITLE
Cant easily see changes to a file in a conversation in subsequent commits

### DISCUSSION THREAD

**Discussion by sparmboy** [community/community#9956] [Pull Requests]
I'm finding this quite an infuriating part of the "conversation" in reviews. This is what Im finding:

* Add conversation to a file in a PR
* Author adds comment of something like "done" and pushes new commit to address the change.
* As reviewer, when reviewing this change, I can see the comment that author has fixed but no way of seeing what the change is. 
 
If I click "view changes" it just displays all the files in the PR not the one with the comment. For large PRs this is a pain as I have to find the file, and then go to the latest commit. But when viewing that file, there is no trace of my original comment / conversation so I cant easily recall where the comment was or what it was about so I have to click back to the conversations list to find it again.

On top of this, when seeing the list of "viewed" files in a PR (i.e. theyre collapsed), there's no indication as to which ones have comments in them. So I have to unclick "viewed" on each of them to see if there was a comment in there that need to be re-reviewed.

Either I'm missing something or this isn't very well thought out?

All I want is for the conversation history to persist in the file across all commits so I can ascertain that the author has indeed made the fix they claim to have

**Comment by Acorath** [comment]
> All I want is for the conversation history to persist in the file across all commits so I can ascertain that the author has indeed made the fix they claim to have

💯 

This topic has been a significant source of frustration for me since my team switched to GitHub. When I take the time to comb through someone else's code and request changes, I need to be able to reference those requests while reviewing their new commits.

**One possible solution that GitHub could implement:** Next to the `View changes` button for a comment in the discussions view, add another button along the lines of `Compare to latest changes`, which would select every commit added to the pull request since the comment was made. If the commented code was altered in any way, then the comments should be displayed on the left pane along with the original changes that were commented on.

**Reply by Kurt-von-Laven** [reply]
> **One possible solution that GitHub could implement:** Next to the `View changes` button for a comment in the discussions view, add another button along the lines of `Compare to latest changes`, which would select every commit added to the pull request since the comment was made. If the commented code was altered in any way, then the comments should be displayed on the left pane along with the original changes that were commented on.

I agree wholeheartedly with the problem described by the OP, but this solution doesn't support the common case of rebasing and force pushing so that comments on each commit are addressed in that commit (rather than a series of "address code review feedback" commits, for example). Such a strategy preserves a much cleaner revision history, which makes tools like `git bisect` and `git blame` far more useful.

**Reply by Acorath** [reply]
@Kurt-von-Laven I don't disagree there, but I'm also not sure what a good solution would look like in that case. Often times the original code that would have been commented on is barely recognizable after a rebase, depending on the scope of changes that occurred upstream. So even if you manage to associate the comments with their respective code, they'll fairly likely lose their context anyway.

But that's more of a team workflow issue I think. Personally, I wouldn't rebase until all of the active change requests have been addressed and resolved, in order to preserve the comments. But there are so many active PRs to my teams repositories that, in the time between opening a PR and getting it peer reviewed and tested, it's all but guaranteed that I'll need to resolve some merge conflicts before I can merge anyway; so I rebase late for that reason.

Curious though, do you have an idea for a solution that would accommodate rebasing?

**Reply by Kurt-von-Laven** [reply]
I imagine GitHub has procrastinated this feature precisely because it is tricky to implement well, and there are many subtle considerations in order to handle a wide variety of workflows well. Your workflow makes a lot of sense to me for your situation! According to https://github.community/t/review-comments-not-shown-in-files-changed-if-a-commit-affects-the-comments-associated-line/1867, BitBucket, GitLab, and Gitter all implement great solutions to this exact same problem, and the latter is [open-source](https://gitlab.com/gitterHQ/webapp) and MIT-licensed. I would encourage GitHub to start by looking at their competitors who have collectively devoted a great deal thought to this problem.

**Reply by paulius-valiunas** [reply]
@Kurt-von-Laven since Github is now owned by Microsoft, they don't even have to look at their competitors, they have Azure DevOps. I don't understand why they still haven't moved some of that functionality over to GitHub, the ADO pull request experience is so much better that GitHub looks like someone's homework project compared to it.

**Reply by hazz0003** [reply]
My company has just recently switched from Azure DevOps Git Repos to GitHub.  I have to reluctantly acknowledge that the Microsoft Azure offering has a significantly better experience here.  They do this.. and do it fairly pretty well - it updates nearly live as new comments and pushes to the PR branch come in.

**Reply by dcumings** [reply]
We switched from bitbucket.  They had 4 ways of helping solve this
1. Conversations show in side tab and take you to the current version of the file and directly to the comment in the current context if possible.
2. When possible the comment would still remain exactly where it was left even if the file changed around it.
3. If the file changed to much the comment would be listed at the top of the file with the ability to pop up the comment and the context of where the comment was at the time the comment was made.
4. If the file was gone you could always pull up the comment and context as a pop up from the side menu.

There where a few other enhancements made that made Bitbucket PR a better experience overall but they are not related to this topic.

![image](https://user-images.githubusercontent.com/2699266/189704181-ab87dfa8-8d8d-4898-9c1c-ff08a2ac0548.png)
![image](https://user-images.githubusercontent.com/2699266/189703961-71285bf5-9826-49ec-9c43-617482bbda81.png)

**Reply by dcumings** [reply]
My current workaround solution is to just have the conversation open in a separate window and manually search for the file.
![image](https://user-images.githubusercontent.com/2699266/190207416-3e09a98e-b993-4f50-847d-f1f6c7abdaf8.png)



The problem is not as bad on small PR's but managing this on large PR's is a nightmare

**Reply by saleazer** [reply]
This is what I have been doing as well, and I miss ADO so bad already because of this..

**Reply by flobacher** [reply]
same here.. I thought ADO is the legacy product and github sets the gold-standard here, but in a big PR its really hard to keep an overview, if the reviewrequester really implemented suggestions or not.
please fix this 🙏

**Comment by xmo-odoo** [comment]
> Either I'm missing something or this isn't very well thought out?

It's not, but FWIW in the regular timeline when the developer force-pushes (replaces) the commits of the PR there's a note saying:

> @user force-pushed the <branchname> branch from <commit1> to <commit2>

Well it's completely invisible because it was specifically styled like regular text but the words "force-pushed" are actually hyperlinked to the *textual diff* between commit1 and commit2 (unlike the link from the notifications which links to nothing).

That aside, if you have the possibility you may want to take a gander at reviewable.io which has a much more complete and flexible flow especially for reviews which are not trivial. Amonst other things they have what are essentially "review points" orthogonal from the commits and PR updates. It's a layer on top of github (and in fact works only with github) so it's not something you have to change everything over.

**Reply by paulius-valiunas** [reply]
Looks nice, except the UI was drawn by a 4th grader with a crayon. But the real deal-breaker is that for public repos, the whole PR flow is hidden away on an external service. It would only be usable if the comments created there could be viewed and replied to natively on Github too, and vice versa. That way anyone browsing Github would be able to view and participate in the discussion related to every PR without having to sign up for an external service.

**Reply by Kurt-von-Laven** [reply]
I wonder if it would be possible to write a GitHub Action that automatically links to the correct external review from GitHub.

**Reply by xmo-odoo** [reply]
@Kurt-von-Laven Reviewable adds a big button to the PRs it starts handling. That's actually why I couldn't use it myself: I just wanted to use it under the covers to make reviews easier on large PRs where github is *really* bad, but the big button was a bit too visible.

**Comment by jababda** [comment]
Does anyone know if there's any planned work at github for this? This seems like a pretty core feature that is frankly a huge step down from bitbucket's or Azure devop's PR experience.

**Comment by aramyesildeniz** [comment]
It is possible to jump through the conversations and comments within the `Files changed` tab, which allows you to quickly see which file and line is affected and eventually got changed.

Go to: `Pull requests -> Files changed `

Find there the submenu: `Conversations -> Jump to conversation`

**Reply by paulius-valiunas** [reply]
that doesn't show the conversations whose context lines have been changed, does it?

**Reply by asilverman-stripe** [reply]
Nope; if you jump to a resolved conversation, it only shows the commit where the conversation was made, not any of the changes made afterward. So you can't verify the changes against the suggestion without going back and forth or opening multiple windows.

**Comment by clemwo** [comment]
This is frustrating coming from GitLab and so weird that this doesn't work well. Hopefully there will be an improvement coming soon

**Comment by robertodraGit** [comment]
Maybe a couple of upvotes make them work on this...

**Comment by ben-polinsky** [comment]
This is a source of significant frustration. Any feedback from GH would be massively appreciated.

**Comment by klinamen** [comment]
Coming from Gitlab, I thought it was me who could not understand how to do such a basic thing while reviewing. Then I bumped into this discussion. IMHO current behavior makes the whole review process an unmotivated pain, and it is a valid reason to hop to another tool.

**Reply by sadikerd** [reply]
Same here! "Coming from Gitlab"

**Comment by Franklyn-R-Silva** [comment]
It sounds like you're encountering some challenges while using a version control system, such as Git, in a collaborative environment for code reviews. Based on your description, it seems like you're looking for a way to have a more streamlined and efficient process for reviewing changes and keeping track of conversations in pull requests (PRs). While I can't provide real-time updates on the latest developments, I can offer some general suggestions that might help address the issues you're facing:

1. **Review Tools**: Make sure you're using the appropriate tools or platforms for code reviews. Platforms like GitHub, GitLab, and Bitbucket provide specific features for managing conversations and reviewing changes in PRs.

2. **Inline Comments**: Rather than just a general "done" comment, encourage authors to use inline comments within the code changes they make. This way, reviewers can easily see the specific changes made in response to their comments.

3. **Diff Views**: Most code review platforms provide a "diff" view that shows changes side-by-side with the original code. This view should highlight what lines were added, modified, or removed. Make sure you're using this feature to track changes more effectively.

4. **Conversation Context**: Some platforms allow you to reply to specific comments within the conversation. This can help keep discussions tied to specific lines of code and provide context to both the reviewer and the author.

5. **Commit Messages**: Encourage authors to use meaningful commit messages that reference the issue or comment being addressed. This can help reviewers understand the purpose of the commit without having to dive into the code changes immediately.

6. **Collaborative Workflow**: Consider adopting a more collaborative workflow, where discussions and decisions are made during the code review process. This could reduce the need for multiple iterations and make it easier to track changes.

7. **Documentation**: If your current platform doesn't offer the features you're looking for, consider creating a standard documentation or process for your team's code review. This can help ensure that everyone follows a consistent approach.

8. **Feedback to Platform**: Many code review platforms actively seek feedback from users to improve their tools. If you find that a particular platform's features are lacking, consider providing feedback or exploring if there have been updates or plugins/extensions that address your concerns.

It's possible that the tools you're using might have improved or introduced new features since my last knowledge update in September 2021. I recommend checking the documentation or support resources of the specific platform you're using to see if there have been updates that address the challenges you've mentioned.

**Reply by klinamen** [reply]
Thank you ChatGPT for your invaluable wisdom!

**Reply by myedibleenso** [reply]
-- brought to you by Open AI

**Comment by carlospnav** [comment]
I came here looking for a solution to exactly this problem since my company migrated from BitBucket to Github. This is quite frustrating. How is it possible that nobody has deemed this issue to be top priority?

We must be missing something, this is such a basic process of reviewing PRs. Either that or we're using a whole different review paradigm than the folks at github are when they're going over their own changes.

**Comment by stephenjason89** [comment]
if github won't support this. I hope refined github will.
Please upvote this
https://github.com/refined-github/refined-github/issues/7048

**Reply by xmo-odoo** [reply]
Refined only does UI tweaks, this is way too involved for that. You need a complete layer of additional metadata over github to fix it, as in [Reviewable](https://reviewable.io/) (which does in fact fix this, amongst others).

**Reply by paulius-valiunas** [reply]
I don't think extra metadata is necessary. Just display all existing comments, regardless of their commit checksum, when showing the latest version of a file. All you need is the file name and line number attached to each comment.

**Comment by Cap-JaTu** [comment]
Anybody have an idea how to download the conversation changes to my own computer for testing? This isn't documented anywhere.

**Reply by paulius-valiunas** [reply]
Do you mean you don't know how to clone a git repository?

**Reply by Cap-JaTu** [reply]
No.

How to have a version with PR contents applied for local build.
Running a clone will give the version without the PR.

**Reply by paulius-valiunas** [reply]
you just checkout the branch that the PR was opened from...

**Reply by Cap-JaTu** [reply]
Yet again, the result will be the branch without the contents of the PR.
I'm supposed to evaluate if a PR is good or not. Doing a build on the changed-but-not-yet-merged code would be a good starting point.

**Reply by Kurt-von-Laven** [reply]
That is what is on the head branch like @paulius-valiunas said.

**Reply by Cap-JaTu** [reply]
If a PR would be in HEAD already, why bother accepting the request?
There has to be a flaw in logic where all the changes in a non-approved pull request will appear in HEAD.

**Reply by xmo-odoo** [reply]
The head *of the PR*, not the head of the target repository. You can clone the repository *the pull request comes from* to check out the branch, or you can use github's [magic pull refs](https://stackoverflow.com/questions/72022248/git-refs-pull-vs-pull) to just check them out.

Now please move this subthread somewhere else, it has nothing to do with the subject.

**Reply by Cap-JaTu** [reply]
Well, maybe this should be documented somewere?

**Reply by paulius-valiunas** [reply]
no need to document this, it's not specific to github, it's just git 101. A "pull request" is a request to merge one branch into the other. If you want to see what is on the branch, just checkout the branch. There is no such thing as "contents of a PR" - there are only contents of a branch that someone wants to merge into another branch (the "other branch" is usually master or main).

**Comment by Cap-JaTu** [comment]
Actually, it's not 101. If I have to ask, I don't know this. Having a few words in any documentation wouldn't hurt.

**Reply by paulius-valiunas** [reply]
I understand you don't know this, but all I can suggest is look up some git tutorials. This is really the basics of the basics.

**Reply by Cap-JaTu** [reply]
This is off-topic, we need to wrap this up.
Trust me, I have looked. It is NOT basics. Any basic material covers only basic cases. Not PR-handling.

**Reply by paulius-valiunas** [reply]
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests

> Pull requests let you tell others about changes you've pushed to a branch in a repository on GitHub. Once a pull request is opened, you can discuss and review the potential changes with collaborators and add follow-up commits before your changes are merged into the base branch.

I think this makes it perfectly clear that the changes discussed in a PR reside on a branch. If you still need additional clarification after reading that paragraph (which is literally the first paragraph on the page about PRs), you can continue reading here:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-branches
... which in turn has a link to this:
https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell

I think the documentation is more than sufficient.

**Reply by Cap-JaTu** [reply]
Thank you for the completely useless links. I'm concluding this fruitless ping-pong and moving forward to find actually usable documentation on how to work with evaluating pull requests. Theory on PRs and documentation on branches don't do the trick here.

**Reply by paulius-valiunas** [reply]
🤦‍♂️

**Reply by robertodraGit** [reply]
@paulius-valiunas I guess that @Cap-JaTu want to know if is it possible to review changes on a branch before GitHub auto-closes conversations. When you leave a comment on someone pr's in GitHub, if he push some code that overwrites the lines of the commented part, GitHub sometimes auto-closes conversations and you lose the content of that messages (actually you can go back commit by commit and view the changes, but in my opinion this option isn't fast and stress-less enough).

Maybe I didnt understand a thing about your convo

**Reply by paulius-valiunas** [reply]
> @paulius-valiunas I guess that @Cap-JaTu want to know if is it possible to review changes on a branch before GitHub auto-closes conversations. When you leave a comment on someone pr's in GitHub, if he push some code that overwrites the lines of the commented part, GitHub sometimes auto-closes conversations and you lose the content of that messages (actually you can go back commit by commit and view the changes, but in my opinion this option isn't fast and stress-less enough).
> 
> Maybe I didnt understand a thing about your convo

That's an entirely different interpretation I haven't considered. @Cap-JaTu, did you want to download the _comments_ instead of source code? In that case I can't really imagine in what format they would be stored for download, and I don't see the point either because comments are never deleted (unless you delete them explicitly).

@robertodraGit Github doesn't auto-close conversations, it just stops displaying them in the files tab. You can still see them in the conversation tab.

**Comment by didmar** [comment]
Bump, this is a really big UX issue, enough to consider moving to Gitlab or Bitbucket which do manage this

**Comment by Caitlin-Syntax** [comment]
Adding my bump here in addition to discussions [17481](https://github.com/orgs/community/discussions/17481) and [23138](https://github.com/orgs/community/discussions/23138). This issue continues to frustrate me regularly.

**Comment by Tommuh** [comment]
Doing daily 10+ reviews. I cannot remember what I have said, it is difficult to verify if the asked changes are actually applied.

**Comment by IRHayfield** [comment]
I agree that GitHub should change this and make outdated comments an optional feature.
As a workaround I created [outdated-parrot](https://github.com/IRHayfield/outdated-parrot) an action which adds all open outdated comment threads as annotations so they are displayed in the files changed view.
This is not an perfect solution, but does allow users to review comments against the correct line with the addition of some links that users may find useful.

**Comment by Caitlin-Syntax** [comment]
This issue is now documented in at least four places with no real official response.
- https://github.com/orgs/community/discussions/9956 (this thread)
- https://github.com/orgs/community/discussions/17481
- https://github.com/orgs/community/discussions/23138
- https://github.com/orgs/community/discussions/110796 (feature request)

**Comment by CoolGoose** [comment]
I  guess we need to bump all a bit more, it's the most annoying thing coming over gitlab, initial context of seeing how the new change 'fixes it' it's important.

**Comment by tabiStein** [comment]
Chiming in to add to the numbers. This is horrible. Makes me really miss SmartBear's Collaborator tool, the gold-standard "ideal" code reviewing experience (I think) -- except that it's a separate tool on top of version control. Would be nice to have such a great experience integrated into Github PRs.

**Comment by szevzol** [comment]
I came here to bump. Facing this issue in a large PR is a a nightmare.

**Comment by paulius-valiunas** [comment]
Will someone please react to this in any way?

**Comment by victorn-op** [comment]
Coming from Azure DevOps and really frustrated with this behavior. Even more frustrated now knowing every other version management platform works correctly except for GitHub.

**Reply by rlopez-syndigo** [reply]
I second that.

**Comment by Caitlin-Syntax** [comment]
New year, new bump. This issue, which has been getting upvotes and comments across at least 4 threads since March 2019, still has no official response. My previous comment links to the threads documenting this issue and requesting UX changes. Please, please fix this, GitHub!

**Comment by fremis-LF** [comment]
Glad to know I'm not crazy for thinking this should be a feature. Juggling multiple reviews in parallel is made really difficult by this!

**Comment by Caitlin-Syntax** [comment]
Another six months, another new bump. Still awaiting any functional response to any of the (at least) [four discussions](https://github.com/orgs/community/discussions/9956#discussioncomment-10593019) about this issue over six years. This is still making my PR review process profoundly annoying.

**Reply by ben-polinsky** [reply]
Maybe we can rephrase the issue to use copilot, then they'll get on it :)

**Comment by rlopez-syndigo** [comment]
This is absolutely terrible.

**Comment by Kusnierz-Maciej** [comment]
Are there any plans for fixing this?

**Comment by rickhcchan** [comment]
On VS Code, after you signed in GitHub, there should be a "Comments" view, and we will see all comments there (marked as outdated)
You then right click and select "Diff comment with HEAD". This should show you the change side by side

**Comment by nbedard-sherweb** [comment]
Just switched from Azure Dev Ops to GitHub. This thread is so accurate. 

Are there any plans for fixing this?

**Reply by deyanp** [reply]
> Just switched from Azure Dev Ops to GitHub. This thread is so accurate.

Same here ...

**Comment by Scott-NW** [comment]
We've just moved from DevOps and also struggling with this!

**Comment by chiaraatbolt** [comment]
+1 to the issue, I thought it was just me not understanding the interface (after coming from Gitlab), but now I realised that this very basic feature is missing? :(

**Comment by adamkarancsi** [comment]
This is a huge pain point for us as well.

**Comment by stephen-trfg** [comment]
+1 in 2026. Coming from DevOps, where this is handled well, it’s surprising to see such a limiting design choice.

**Comment by alexander-furmenkov-altium** [comment]
Could this finally be fixed?

**Comment by Caitlin-Syntax** [comment]
I'd recommend we all go boost the open feature request listed below. This issue has been documented in (at least) the four threads below since 2019. I wonder if we're going to make it to a decade with no fix...

- https://github.com/orgs/community/discussions/9956 (this thread)
- https://github.com/orgs/community/discussions/17481
- https://github.com/orgs/community/discussions/23138
- https://github.com/orgs/community/discussions/110796 (feature request)

**Comment by hari90** [comment]
+1 This is a huge setback for our team coming especially given we are moving over from an older system that handled it better.
